[replaced] Zero-Rotating Marker/Popup/Tooltip Pane

sisou opened the pull request:
In reference to the comment at https://github.com/fnicollet/Leaflet/issues/8#issuecomment-249875366, that the markers, popups and tooltips should not be rotated back and forth, but instead be rotated _zero_ times, I started this branch to try to accomplish that.

I split the map panes in two groups, leaving the tilePane and the overlayPane within the rotatePane, but creating a new norotatePane for the three other panes.

I adapted a bit of the positioning logic, but I am now stuck.

What works:
Panning, zooming, initial rotating

What does not work:
Rotating, then panning, then rotating (or zooming). The origin of the second rotate/zooming action is wrong for the elements in the no-rotate pane. I don't know how to fix it currently and am hoping for the input of experienced Leaflet devs here ;-) I'm guessing I need to reset the origin somewhere, somehow...?

You can try out the changes and problems at
https://rawgit.com/sisou/Leaflet/zero-rotate/debug/rotate/rotate.html
https://rawgit.com/sisou/Leaflet/zero-rotate/debug/rotate/rotate-and-drag.html

Commit:
First step towards a zero-rotating marker/popup/tooltip pane

Commit:
Set range inputs to 0 for rotate examples

fnicollet commented:
Hello @sisou and thanks for that PR!
Isolating the markers / popups to a non-rotating pan is a good idea, instead of counter-rotating the objects. 
The fact that it is off when rotating then paning then rotating probably comes from the mapPivot calculation, which will need to take into account the pane container offset or something. I lost a few nights over issues like this and I am not sure how to fix it properly, but there is a way

i'll try your branch when I have a little time!
Fabien

Commit:
Successfully transition rotatePane transform to same way like markers

Commit:
tmp

Commit:
progress!!! =)

Commit:
Fix DivOverlays

Commit:
Change back unnecessary changes

Commit:
blub

sisou committed:
Merge pull request #1 from sisou/temp

Fix rotation after rotated panning

sisou commented:
**Progress**

I managed to make the markers and divOverlays behave nicely.

Next up: zooming positioning of divOverlays after a rotated panning... -.-

Commit:
Reduce diff

Commit:
Fix animated zooming for DivOverlays

sisou commented:
**Progress**

Animated zooming of popups and tooltips (I hope, did not test tooltips) works.

Next up:

sisou commented:
Ups, wrong button :D

Commit:
Fix autoPan for popups when rotated

Commit:
Simplify diff view

Commit:
Clean-up

Commit:
Fix z-index positioning of markers

Because the z-index was not rounded (rotated positions are mostly floats), the style got not accepted by the browser.

Commit:
Add missing tooltip pane & clean-up 2

Commit:
Fix marker dragging when rotated

Commit:
Remove markerRotate option

This option/functionality is a non-trivial implementation and thus currently discarded.

Commit:
Add tooltip to rotate-and-drag demo

Commit:
Fix tooltip positioning

Commit:
DRY up tranformation calculations

Commit:
Round the position of everyting on the norotatePane

Since all elements are still rendered with float positions because of rotation calculations, this is needed to make all icons/popups/tooltips tuly sharp at all times.

Commit:
Rename and add comments to DRY functions

sisou commented:
**Completed**

Hi Fabien,

I think I finished this! I tested everything (also with rotation off) and it works great. Please have a look through and see if you find any problems =)

(Before merging, I think I should squash the commits to make them look nicer :D)

sisou commented:
This PR has been superseeded by #21.
